Objects/perf_trampoline.c

```diff
@@ -0,0 +1,465 @@
+        if (new_code_arena() < 0) {
+            return NULL;
+        }
+    }
+    assert(code_arena->size_left <= code_arena->size);
+    return code_arena_new_code(code_arena);
+}
+
+static PyObject *
+py_trampoline_evaluator(PyThreadState *ts, _PyInterpreterFrame *frame,
+                        int throw)
+{
+    if (perf_status == PERF_STATUS_FAILED ||
+        perf_status == PERF_STATUS_NO_INIT) {
+        goto default_eval;
+    }
+    PyCodeObject *co = frame->f_code;
+    py_trampoline f = NULL;
+    _PyCode_GetExtra((PyObject *)co, extra_code_index, (void **)&f);
+    if (f == NULL) {
+        // This is the first time we see this code object so we need
+        // to compile a trampoline for it.
+        if (extra_code_index == -1) {
+            extra_code_index = _PyEval_RequestCodeExtraIndex(NULL);
+        }
+        py_trampoline new_trampoline = compile_trampoline();
+        if (new_trampoline == NULL) {
+            goto default_eval;
+        }
+        trampoline_api.write_state(trampoline_api.state, new_trampoline,
+                                   code_arena->code_size, co);
+        _PyCode_SetExtra((PyObject *)co, extra_code_index,
+                         (void *)new_trampoline);
+        f = new_trampoline;
+    }
+    assert(f != NULL);
+    return f(_PyEval_EvalFrameDefault, ts, frame, throw);
+default_eval:
+    // Something failed, fall back to the default evaluator.
+    return _PyEval_EvalFrameDefault(ts, frame, throw);
+}
+#endif  // _PY_HAVE_PERF_TRAMPOLINE
+
+int
+_PyIsPerfTrampolineActive(void)
+{
+#ifdef _PY_HAVE_PERF_TRAMPOLINE
+    PyThreadState *tstate = _PyThreadState_GET();
+    return tstate->interp->eval_frame == py_trampoline_evaluator;
+#endif
+    return 0;
+}
+
+int
+_PyPerfTrampoline_SetCallbacks(trampoline_state_init init_state,
+                               trampoline_state_write write_state,
+                               trampoline_state_free free_state)
+{
+#ifdef _PY_HAVE_PERF_TRAMPOLINE
+    if (trampoline_api.state) {
+        Py_FatalError("Trampoline state already initialized");
+        return -1;
+    }
+    trampoline_api.init_state = init_state;
+    trampoline_api.write_state = write_state;
+    trampoline_api.free_state = free_state;
+    trampoline_api.state = NULL;
+    perf_status = PERF_STATUS_OK;
+#endif
+    return 0;
+}
+
+int
+_PyPerfTrampoline_Init(int activate)
+{
+#ifdef _PY_HAVE_PERF_TRAMPOLINE
+    PyThreadState *tstate = _PyThreadState_GET();
+    if (tstate->interp->eval_frame &&
+        tstate->interp->eval_frame != py_trampoline_evaluator) {
+        PyErr_SetString(PyExc_RuntimeError,
+                        "Trampoline cannot be initialized as a custom eval "
+                        "frame is already present");
+        return -1;
+    }
+    if (!activate) {
+        tstate->interp->eval_frame = NULL;
+    }
+    else {
+        tstate->interp->eval_frame = py_trampoline_evaluator;
+        if (new_code_arena() < 0) {
```

Comment:
if `new_code_arena` fails then it should fallback to eval frame default not py_trampoline_evaluator.